Base implementation of Dynamix, Tim

molenzwiebel opened the pull request:
This PR contains the following changes:

- A new metalang, tim, which is the result of running a Dynamix specification on an input AST. This is a CPS-based intermediate representation language. The tim metalang includes a simple Stratego interpreter, but the intention is that this is replaced with a more performant optimizing compiler in a follow-up master thesis.
- The dynamix metalang has been split into dynamix (grammar, static analysis, "compilation"/spec merging) and dynamix_runtime (stratego interpreter for running compiled/merged specifications).
- Enable multi-file analysis for dynamix, and support running the statix spec inside the concurrent interpreter.
- Tasks/menus for compiling the dynamix specification in a project to an ATerm representation, and using the dynamix_runtime for executing the spec on an input AST.
- All languages (tim/dynamix/dynamix_runtime) have been namespaced into folders, their sort/constructor names prefixed (T/M respectively) and stratego rules prefixed (tim-, dx- respectively).
- The `tim`/`dynamix`/`dynamix_runtime` spoofax2 projects use source imports to depend on each other.
- Actual signature generation for dynamix is implemented, instead of simple stubs.

Notes/issues that would be nice for you to take a look at:

- The `ReadDynamixSpecTaskDef` is possibly a bit over the top and could be changed to just passing a `ResourcePath`, and having the reading be done in the dynamix runtime instead, but I wasn't entirely sure if that would work properly with PIE.
- Menu items are included for running the Dynamix specification. However, the continuous versions of these seem to not work. Possibly due to the fact that the output is written to a file.
- Embedding dynamix_runtime should be complete (largely copied from the rv32im code), but I'm not entirely certain if is correct. I am not sure whether a standalone eclipse build will work.
- Passing options to the DynamixRuntime is currently done through `createDynamixRuntimeConfig` in `ExecuteDynamixSpecificationTaskDef.java.mustache`. This works, but there might be a more idiomatic approach?
- The `dynamix_runtime` "language" is currently in the metalang folder, but it might be more appropriate to put it elsewhere (next to rv32im?).
- The `tim` stratego interpreter is currently bundled with the metalang (and not split into `tim_runtime` or something similar). Currently done because it works and as mentioned this shouldn't be the long-term solution for running tim files.
- Source imports for `ESV` files don't seem to work. Try uncommenting the `tim/colors` import in the dynamix Main.esv.
- (Spoofax 2) versions of all the languages are set to `0.1.0-SNAPSHOT`. Is that fine?
- All stratego files have commented out type signatures for str2. Ideally the files are ported over the str2, but I had some issues with getting Stratego 2 to work properly in Spoofax 2 so we're back on the old stratego for now.

Other notes/TODOs:

- Spoofax 3 doesn't yet support configuring the Statix solver mode. @AZWN, I see that #108 is open for that?
- Current statix spec becomes very slow very fast, big culprit here is the checking for overlaps in rule patterns. This is partially on my end, and partially something which might be due to an issue in the Statix solver?
- The Statix spec seems to work for the concurrent solver (tested in Spoofax 2), but the concurrent solver is significantly slower. The incremental solver seems to fail. I sent a bundled version to @AZWN, hopefully he has some time to take a look at some point.

molenzwiebel committed:
Adapter/task compilation for Dynamix, depend on runtime package

molenzwiebel committed:
Namespace Dynamix files, cleanup

molenzwiebel committed:
Finalize signature generation for Dynamix

molenzwiebel committed:
Add Tim language implementation and runtime

molenzwiebel committed:
Add Dynamix Runtime implementation

molenzwiebel committed:
Add support for constraint-analyzer property lookup and globals in Dynamix

molenzwiebel committed:
Allow configuration of Dynamix entry-point rule

molenzwiebel committed:
Minor fixes, include source location in rules for better stack traces

molenzwiebel committed:
Consult rules in order of specificity during Dynamix evaluation

molenzwiebel committed:
Check for overlapping patterns in Statix, add support for concurrent solver

molenzwiebel committed:
Add (untested) #array-write primitive to Tim

molenzwiebel committed:
Merge from upstream

molenzwiebel commented on lwb/metalang/dynamix_runtime/dynamix_runtime/src/main/java/mb/dynamix_runtime/task/DynamixRuntimePrettyPrint.java:
```diff
@@ -0,0 +1,45 @@
+
+@DynamixRuntimeScope
+public class DynamixRuntimePrettyPrint implements TaskDef<IStrategoTerm, Result<String, ?>> {
+    private final Provider<StrategoRuntime> strategoRuntimeProvider;
```
Using `DynamixRuntimeGetStrategoRuntimeProvider` here results in an exception in PIE on every build after the first, due to reasons I don't fully understand. For now this is acceptable, since the stratego ctree/jar of the runtime will be bundled with Eclipse anyway, but we should likely get a proper dependency on it.

Gohla replied:
Can you add a TODO in the code here that we should switch to `DynamixRuntimeGetStrategoRuntimeProvider` if possible and the reason this was not done?

molenzwiebel committed:
Attempt at more efficiently comparing patterns in Statix

Gohla commented on lwb/metalang/dynamix_runtime/dynamix_runtime/src/main/java/mb/dynamix_runtime/ReadDynamixSpecTaskDef.java:
```diff
@@ -0,0 +1,55 @@
+                specAst = readTerm(inputStream);
+            }
+            return Result.ofOk(specAst);
+        } catch(Exception ex) {
```
Catching all `Exception`s is bad practice, as it also catches all `RuntimeException`s indicating bugs such as `NullPointerException` which should just crash the build, as well as catching `InterruptedException`s which should cancel the build, etc.

Can you catch a more specific exception? If not, can you rethrow `RuntimeException` and `InterruptedException`?

Gohla commented on lwb/metalang/dynamix_runtime/dynamix_runtime/src/main/java/mb/dynamix_runtime/ReadDynamixSpecTaskDef.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            return reader.parseFromStream(stream);
+        } catch(IOException e) {
+            throw new IllegalStateException("Loading ATerm from stream failed unexpectedly", e);
```
Can the `IOException` be handled in a nicer way? Either return the exception as part of the `Result` of the task, or just let the `IOException` bubble up?

Gohla reviewed (changes requested):
Very cool, and also very quickly done!

Are there any examples and/or regression tests for dynamix, to ensure that dynamix will not suddenly break? If not, it would be good to at least have some simple smoke tests.

Answers to general questions:

> The ReadDynamixSpecTaskDef is possibly a bit over the top and could be changed to just passing a ResourcePath, and having the reading be done in the dynamix runtime instead, but I wasn't entirely sure if that would work properly with PIE.

Maybe the reading can be done by the dynamix runtime instead of outputting an AST from a task (since reading is probably so fast that caching may not be useful?). However, for incrementality you must still tell PIE about the read (require) dependency, otherwise changing the specification will not result in rerunning the dynamix runtime.

> Menu items are included for running the Dynamix specification. However, the continuous versions of these seem to not work. Possibly due to the fact that the output is written to a file.

Hmm, this could be a bug, can you make a bug report so I can look into this sometime?

> Embedding dynamix_runtime should be complete (largely copied from the rv32im code), but I'm not entirely certain if is correct. I am not sure whether a standalone eclipse build will work.

From a glance it looks ok, but can you test this? If you run `gradle buildSpoofax3LwbEclipseInstallation` in devenv it will create a standalone installation somewhere in `spoofax.pie/lwb.distrib/spoofax.lwb.eclipse.repository/build/` (either in `eclipse-<os>-<arch>` or in the `distributions` directory. If there are problems we can look into it.

> Passing options to the DynamixRuntime is currently done through createDynamixRuntimeConfig in ExecuteDynamixSpecificationTaskDef.java.mustache. This works, but there might be a more idiomatic approach?

That's what we do in other places too, it's easy and works. The downside is that changing configuration requires recompilation of the Java classes. You could also write the options to a [Java properties file](https://en.wikipedia.org/wiki/.properties) during language compilation (be sure to use `mb.common.util.Properties#storeWithoutDate` to ensure no date is written to make it deterministic) and require and read that in the task.

> The dynamix_runtime "language" is currently in the metalang folder, but it might be more appropriate to put it elsewhere (next to rv32im?).

It needs to be part of the `lwb` composite build dependency wise, so it should stay there. Maybe there could be another `metaruntime` directory for projects like that.

> The tim stratego interpreter is currently bundled with the metalang (and not split into tim_runtime or something similar). Currently done because it works and as mentioned this shouldn't be the long-term solution for running tim files.

I guess that is fine.

> Source imports for ESV files don't seem to work. Try uncommenting the tim/colors import in the dynamix Main.esv.

Can you make a bug report for this as well, so I can have a look at it sometime?

> (Spoofax 2) versions of all the languages are set to 0.1.0-SNAPSHOT. Is that fine?

I think the Spoofax 2 Gradle plugin will override that with the version from Gradle, so that should be fine.

> All stratego files have commented out type signatures for str2. Ideally the files are ported over the str2, but I had some issues with getting Stratego 2 to work properly in Spoofax 2 so we're back on the old stratego for now.

That is too bad, but fine solution for now. Did you contact Jeff or make a bug report?

molenzwiebel committed:
Add some more tim primitives, make injected lists more ergonomic in Dynamix statics

molenzwiebel committed:
Make TFuns first-class values, do not generate signatures for bracket productions

molenzwiebel committed:
Extract Tim runtime into separate project, depend on project

molenzwiebel committed:
Reorganize automatically generated Dynamix tasks

molenzwiebel committed:
An attempt at adding some native strategies to Tim

AZWN commented:
To quickly react to your last points involving the concurrent solver: When your specification semantically works with the concurrent solver, and there is no type-directed transformation, I think it is fine to use the traditional solver for now, and switch to the concurrent solver later. I'll add a TODO to that PR.

RvanBaarle committed:
Initial work on a closure conversion strategy

RvanBaarle committed:
Extract closures to outermost level. Fix mutually recursive closures.

RvanBaarle committed:
Add uniquify transform

RvanBaarle committed:
Inlining of functions

RvanBaarle committed:
Support for integer operations

RvanBaarle committed:
Add support for array operations

RvanBaarle committed:
Add support for refs

RvanBaarle committed:
Add support for strings

Breaks printing ints!!!

RvanBaarle committed:
Add partial record support

Arguments to `#record-new(...)` are not yet supported

RvanBaarle committed:
Support for arguments to `#record-new`

RvanBaarle committed:
Add closure type to tim

It functions like an array, but is read only. `closure-offset` allows referencing part of a closure.

RvanBaarle committed:
Allow for negative indices in closure reads/offsets

RvanBaarle committed:
Implement compilation of new closure primops

RvanBaarle committed:
Combine closures for each fix expression

RvanBaarle committed:
WIP for closure conversion improvements

RvanBaarle committed:
Creation of project for LLVM IR

RvanBaarle committed:
Initial version of LLVM syntax

RvanBaarle committed:
Minor syntax fixes

RvanBaarle committed:
Move around files and minor syntax changes

RvanBaarle committed:
Get rid of all string interpolation

The time for ASTs has come

RvanBaarle committed:
C runtime library for Tim

Should probably move it elsewhere in the future

RvanBaarle committed:
Differentiate between printing ints and strings

RvanBaarle committed:
Fix statix type of string literal

RvanBaarle committed:
Add array tail, length and concat operations

RvanBaarle committed:
Minor fixes to syntax

RvanBaarle committed:
Implement the last missing features

Mostly being string operations

RvanBaarle committed:
Fix non-escaping closure optimizations

RvanBaarle committed:
Update compiler flags to emit bitcode

RvanBaarle committed:
Use static closure address if known at compile time

RvanBaarle committed:
Let binding inlining

RvanBaarle committed:
Extend LLVM syntax for GC support

RvanBaarle committed:
WIP in explicating types further

RvanBaarle committed:
More explicit typing

RvanBaarle committed:
Explicitly typed, and progress towards GC

RvanBaarle committed:
quick fixes to correctly compile

RvanBaarle committed:
Change what gets logged

base pointer printing

RvanBaarle committed:
Fix incorrect base pointer

RvanBaarle committed:
Arrays start at one, fix tailcall issue

RvanBaarle committed:
Array type consistency

RvanBaarle committed:
Add gc-leaf-function to function applications

Needed in case the called function can't be determined during compile time (closures)

RvanBaarle committed:
Garbage collector

Currently doesn't yet traverse through objects

RvanBaarle committed:
Remove all unused includes

RvanBaarle committed:
Scan heap for pointers as well

RvanBaarle committed:
Support for finalizers

Called when an object is no longer in accessible memory

RvanBaarle committed:
Add [[maybe_unused]] attributes

RvanBaarle committed:
Fix array and string operations using old references

RvanBaarle committed:
Garbage collection for closures

RvanBaarle committed:
Add LLVM IR to Spoofax LWB

Compiling to LLVM in Spoofax 3 is not yet functional

RvanBaarle committed:
Quick fix for emitting LLVM from Spoofax 3

RvanBaarle committed:
Fix a couple of bugs

* Records were handled incorrectly by gc
* Function type was not converted to closure type everywhere just yet

RvanBaarle committed:
Change stratego provider to ctree in hopes of harmony between spoofax 2&3

RvanBaarle committed:
Add explicit type annotations in preparation of more efficient types

RvanBaarle committed:
Add type annotations to functions as well

RvanBaarle committed:
Integer boxing

RvanBaarle committed:
Bitfield encoding and flag-related bugfixes

WIP better closure typing

RvanBaarle committed:
Typed closures

RvanBaarle committed:
Struct type added

RvanBaarle committed:
Add typing additions to dynamix

RvanBaarle committed:
Implement tuples and fixes to types

RvanBaarle committed:
Rename holy to pluggable

RvanBaarle committed:
Refactor locations of ast.str, type.str and string.str

RvanBaarle committed:
Struct inheritance

RvanBaarle committed:
Small fixes to Tim

RvanBaarle committed:
Improve performance issue, change names to match those used in Thijs' thesis

RvanBaarle committed:
Fix closure conversion edgecase

In some cases free variables might be added after processing a function

RvanBaarle committed:
Fix recursive types and possibility for gc to encounter invalid pointers

RvanBaarle committed:
Fix recursive types better

RvanBaarle committed:
Revert changes to tim type system

RvanBaarle committed:
Have structs contain scopes

RvanBaarle committed:
Adding str-neq condition, slight LLVM syntax fixes, some other small bugfixes

AZWN commented:
Exactly 11 months ago, I posted a comment about enabling the concurrent solver. I don't remember the discussion fully anymore, but the referred PR has landed, so technically, enabling the concurrent solver might be possible.
